src/gun.erl

```diff
@@ -1233,15 +1233,23 @@ connected(internal, {connected, Socket, NewProtocol},
 		State0=#state{owner=Owner, opts=Opts, transport=Transport}) ->
 	{Protocol, ProtoOpts} = gun_protocols:handler_and_opts(NewProtocol, Opts),
 	{StateName, ProtoState} = Protocol:init(Owner, Socket, Transport, ProtoOpts),
+	State1 = State0#state{socket=Socket, protocol=Protocol, protocol_state=ProtoState},
```

Comment:
I'm not happy with this part, `StateName` or `{error,_}` return value. It's probably made my opinion worse that it's not handled in many parts of the code, though. I wonder if there's a better way to handle cases where we have to send something immediately. I haven't had time to think much about it today unfortunately.

Reply:
Is it specifically the return value of `Protocol:init/4` you're concerned with?

I don't know how to better represent it, bit I think init needs to be able to return an error. The related the todos can probably be handled by disconnecting as we do in `commands/2` ([gun.erl:1625)[https://github.com/ninenines/gun/blob/2.0.0-rc.2/src/gun.erl#L1625]) WDYT?

```Erlang
commands([Error={error, _}|_], State) ->
	disconnect(State, Error);
```

Reply:
I need to read more of the Gun code to answer, maybe reorganize a few things. Might take a while as I got a RabbitMQ bug to investigate.

Reply:
OK first I note that the code in this and the switch_protocol command are very similar. So we should work toward making them share the same code. The difference is that one sends a message, the other triggers an event. The shared code should start at Proto:init until the end of the clause. The message/event can be sent before returning next_state instead of where it is sent now (it could be a function that receives the updated State and returns a final State variable).

Perhaps this could be done as an intermediate state (or two) with a forced event that contains the function or similar.

The init functions should either return `{ok, StateName, ProtoState}` or `{error, Reason}`. Then we can match on THAT rather than on `StateName` and make the code more obvious. And later we could refactor to make use of `maybe` (especially for calls in protocols). The rest of the changes should be updated to have ok/error tuples (I see some `State | {error, _}` in `gun_http2` for example).

For the Proto:init/4 and init/6 we need to figure out how they should be handled. I do not think this should be merged until they have been figured out. I can help figure out what to do for the tunnels (some must bubble up to the real socket disconnecting, others just close a stream with optionally a fake socket).

I believe the first part (refactor) can be done separately. I think it would help also to do this on a per-protocol basis, starting with the simplest. A first easy PR would be this small refactor + easy protocols (raw, Websocket, socks, whatever takes over the entire connection). I think `gun_http` shouldn't be too difficult on top of that. `gun_http2` and `gun_tunnel` will likely need to be figured out together.

Thoughts?

Reply:
Sounds good. I can do these changes.

For TCP errors, it's safest to close the socket IMO, not just the HTTP/2 frame. It's almost impossible to recover from a TCP error if the TCP layer hasn't already solved it.

(FWIW, we are using the simple patch which just sends a `tcp_error` message to self. It works well in practice.)

Reply:
It depends from where the error comes from. If it's from the real socket, then yes, everything should be closed. If it's from a fake socket (tunnels), only that tunnel should be closed. But it also depends on what protocol was used to setup the tunnel, as some don't allow subsequent requests (SOCKS) or parallel requests (HTTP, so it's probably safe to close). But HTTP/2 you can have both a tunnel and normal requests on the same connection so in that case we would only close the tunnel.